Central-Server/API/threads.py

```diff
@@ -0,0 +1,282 @@
+    def add_job(self, config):
+        print(f"added job {config} to queues", flush=True)
+        self.queue.append(config)
+    
+    def add_thread(self, thr):
+        self.spin_up_queue.append(thr)
+    
+    def terminate_all(self):
+        for t in self.threads:
+            if t.is_alive():
+                t.terminate()
+                print(f"terminated thread {t.thread_ID}", flush=True)
+        
+        self.running = False
+        print(f"terminated manager", flush=True)
+        
+    def remove_dead_threads(self):
+        self.threads = [thr for thr in self.threads if thr.is_alive()]
+    
+    def kill_thr(self, ident):
+        # TODO
+        """if 0 <= ident < len(self.threads):
+            self.threads[ident].terminate()
+            del self.threads[ident]
+        else:
+            for i, t in enumerate(self.threads):
+                if ident == t.thread_name:
+                    t.terminate()
+                    del self.threads[i]"""
+
+    def check_jobs(self):
+        jobs_to_add = []
+        if self.last_time == None:
+            try:
+                jobs_to_add = self.get_no_last_time_queue()
+            except Exception as e:
+                print(e, flush=True)
+                # Ignore error :(
+                pass
+        else:
+            try:
+                jobs_to_add = self.last_time_queue()
+            except Exception as e:
+                print(e, flush=True)
+                # Ignore error :(
+                pass
+        # Then add to queue
+        for job in jobs_to_add:
+            print(job, flush=True)
+            job_data = packets.JobData(job[0], pull_target=job[2])
+            self.add_job(packets.SV_JOB(job_data, ""))
+
+    def run(self): 
+        if DEBUG:
+            for _ in range(1000):
+                print("debug")
+                # sleep(0.1)
+        else:
+            i = 0
+
+            # Datastreamer websocket implementation
+            def ws_on_connect(sid, environ):
+                self.add_thread(DatastreamerConnection(sid, websocket_channel.ClientWebsocketConnection(self.ws_thread.socketio_server, sid)))
+            
+            def ws_on_message(sid, data):
+                print(sid,": ",data)
+
+            def ws_on_disconnect(sid):
+                self.kill_thr(sid)
+
+            print("(manager_thread) Creating websocket subthread")
+            self.ws_thread = WebsocketThread(5001)
+            self.ws_thread.setup_callbacks(ws_on_connect, ws_on_message, ws_on_disconnect)
+            self.ws_thread.start()
+
+            while self.running:
+                self.check_jobs()
+                self.remove_dead_threads()
+                self.create_threads()
+                if len(self.queue) > 0:
+                    # Find first open board
+                    for t in self.threads:
+                        if t.is_alive():
+                            if t.can_take_job():
+                                if len(self.queue) > 0:
```

Comment:
isn't this condition alr met when entering the original if statement

Reply:
Which condition are you talking about?

Reply:
Fixed.